compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -376,6 +376,17 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
 
         debug!("is_borrowed_temporary: {:?}", self.drop_ranges.is_borrowed_temporary(expr));
 
+        let ty = self.fcx.typeck_results.borrow().expr_ty_adjusted_opt(expr);
+        let may_need_drop = |ty: Ty<'tcx>| {
+            let ty = self.fcx.resolve_vars_if_possible(ty);
+            // Use this instead of `needs_infer` to avoid penalizing types with lifetimes.
+            // This is just a workaround to avoid ICEs anyway.
+            if ty.has_infer_types_or_consts() {
+                return true;
+            }
+            ty.needs_drop(self.fcx.tcx, self.fcx.param_env)
```

Comment:
Will ICE if there are any region variables, I think. try this :smile_cat: 

```suggestion
            self.fcx.tcx.erase_regions(ty).needs_drop(self.fcx.tcx, self.fcx.param_env)
```